wire-moshi-support/src/main/java/com/squareup/wire/MessageJsonAdapter.java

```diff
@@ -0,0 +1,168 @@
+      fieldBindings.put(binding.name, binding);
+    }
+    this.fieldBindings = unmodifiableMap(fieldBindings);
+  }
+
+  @Override public void toJson(JsonWriter out, @Nullable M message) throws IOException {
+    if (message == null) {
+      out.nullValue();
+      return;
+    }
+
+    out.beginObject();
+    for (FieldBinding<M, B> tagBinding : messageAdapter.fieldBindings().values()) {
+      Object value = tagBinding.get(message);
+      if (value == null) {
+        continue;
+      }
+      out.name(tagBinding.name);
+      emitJson(out, value, tagBinding.singleAdapter(), tagBinding.label);
+    }
+    out.endObject();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void emitJson(JsonWriter out, Object value, ProtoAdapter<?> adapter, Label label)
+      throws IOException {
+    if (adapter == ProtoAdapter.UINT64) {
+      if (label.isRepeated()) {
+        List<Long> longs = (List<Long>) value;
+        out.beginArray();
+        for (int i = 0, count = longs.size(); i < count; i++) {
+          emitUint64(longs.get(i), out);
+        }
+        out.endArray();
+      } else {
+        emitUint64((Long) value, out);
+      }
+    } else {
+      moshi.adapter(Object.class).toJson(out, value);
+    }
+  }
+
+  private void emitUint64(Long value, JsonWriter out) throws IOException {
+    if (value < 0) {
+      BigInteger unsigned = POWER_64.add(BigInteger.valueOf(value));
+      out.value(unsigned);
+    } else {
+      out.value(value);
+    }
+  }
+
+  @Nullable @Override public M fromJson(JsonReader in) throws IOException {
+    if (in.peek() == JsonReader.Token.NULL) {
+      in.nextNull();
+      return null;
+    }
+
+    B builder = messageAdapter.newBuilder();
+
+    in.beginObject();
+    while (in.hasNext()) {
+      String name = in.nextName();
+
+      FieldBinding<M, B> fieldBinding = fieldBindings.get(name);
+      if (fieldBinding == null) {
+        in.skipValue();
+      } else {
+        Object value = parseValue(fieldBinding, in);
+        fieldBinding.set(builder, value);
+      }
+    }
+
+    in.endObject();
+    return builder.build();
+  }
+
+  private Object parseValue(FieldBinding<?, ?> fieldBinding, JsonReader in) throws IOException {
+    ProtoAdapter<?> protoAdapter = fieldBinding.singleAdapter();
+
+    if (protoAdapter == ProtoAdapter.UINT64) {
+      if (fieldBinding.label.isRepeated()) {
+        List<Long> longs = new ArrayList<>();
+        in.beginArray();
+        while (in.hasNext()) {
+          longs.add(consumeUint64(in));
+        }
+        in.endArray();
+
+        return longs;
+      } else {
+        return consumeUint64(in);
+      }
+    }
+
+    Object object = moshi.adapter(Object.class).fromJson(in);
+
+    if (fieldBinding.label.isRepeated()) {
+      if (object == null) {
+        return Collections.emptyList();
+      }
+
+      Class<?> itemType = protoAdapter.javaType;
+
+      return moshi.adapter(Types.newParameterizedType(List.class, itemType)).fromJsonValue(object);
```

Comment:
gotta find a way to do these moshi.adapter lookups eagerly

Reply:
Struggling with this one. Even if I had to target type via `itemType`, I don't know how to avoid a lookup for an adapter through `moshi.adapter()`

Reply:
I changed the logic so I don't do _two_ adapter lookup, but not sure that is enough.